src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,204 @@
+    }
+    // TODO: 142990.
+    // Need sort after update loop ports. There are possibility that all exprs are moved to outer loop.
+    if (!inner_loop_info->get_input_ports().empty() && !inner_loop_info->get_output_ports().empty()) {
+        loop_manager->sort_loop_ports(inner_loop_begin_pos, inner_loop_end_pos, inner_loop_id);
+    }
+}
+
+std::set<ExpressionPtr> ExtractLoopInvariants::get_potential_extractable_exprs(const std::vector<LoopPort>& loop_in_ports) {
+    std::set<ExpressionPtr> expr_set;
+    for (size_t i = 0; i < loop_in_ports.size(); ++i) {
+        expr_set.insert(loop_in_ports[i].expr_port->get_expr());
+    }
+    return expr_set;
+}
+
+bool ExtractLoopInvariants::extract_from_loop(const size_t& inner_loop_id, LinearIR& linear_ir) {
+    const auto& loop_manager = linear_ir.get_loop_manager();
+    bool status = false;
+    bool extraction_completed = false;  // true means all extractable exprs are extracted from this loop
+    const auto& inner_loop_info = loop_manager->get_loop_info<UnifiedLoopInfo>(inner_loop_id);
+    while (!extraction_completed) {
+        const auto& inner_loop_input_ports = inner_loop_info->get_input_ports();
+        const auto& potential_extractable_exprs = get_potential_extractable_exprs(inner_loop_input_ports);
+        bool has_been_extraction = false;
+        for (const auto& port_expr : potential_extractable_exprs) {
+            if (is_extraction_applicable(port_expr, inner_loop_info)) {
+                status = true;
+                LinearIR::constExprIt inner_loop_begin_pos, inner_loop_end_pos;
+                std::tie(inner_loop_begin_pos, inner_loop_end_pos) =
+                    loop_manager->get_loop_bounds(linear_ir, inner_loop_id);
+                // extract scalar on inputs if there are
+                for (size_t i = 0; i < port_expr->get_input_count(); ++i) {
+                    auto parent = port_expr->get_input_port_connector(i)->get_source().get_expr();
+                    if (ov::is_type<snippets::op::Scalar>(parent->get_node())) {
+                        extract_expr(parent, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
+                    }
+                }
+                extract_expr(port_expr, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
+                update_loop_ports(port_expr, loop_manager, inner_loop_id, inner_loop_begin_pos, inner_loop_end_pos);
+                has_been_extraction = true;
+                break;  // extracted and refreshed loop_input_ports. break potential_extractable_exprs loop, and go while() to start again.
+            }
+        }
+        if (inner_loop_input_ports.size() == 0 && inner_loop_info->get_output_ports().size() == 0) {
+            // become a empty(inner_loop_input_ports is ref) loop after extraction, let remove it from loop_manager
+            loop_manager->remove_loop_info(inner_loop_id);
+            extraction_completed = true;
```

Comment:
Why do we the `extraction_completed` variable? You use `break` to get out of the `while` anyways.

Reply:
This break is to get out of for `potential_extractable_exprs`. Let's see [comment ](url)for more details.